Change GetAppendOnlyEntryAuxOids/GetAppendOnlyEntry to macros

huansong opened the pull request:
After moving pg_appendonly tuples into relcache, it seems that we no longer need to incur function calls just for some memory copy/dereference. Change them to macros now.

Per discussion in https://github.com/greenplum-db/gpdb/pull/15027#discussion_r1120460866

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

soumyadeep2007 commented on src/include/catalog/pg_appendonly.h:
```diff
@@ -99,29 +99,39 @@ InsertAppendOnlyEntry(Oid relid,
+/* copy the pg_appendonly entry out */
+#define GetAppendOnlyEntry(rel, aoEntry)  \
+do { \
+	memcpy(aoEntry, ((Relation)rel)->rd_appendonly, APPENDONLY_TUPLE_SIZE); \
```
Its best to enclose every macro arg with `()` every time you reference it in the body, for safe expansion. Please do this for the other one as well.

huansong replied:
Done!

soumyadeep2007 commented on src/include/catalog/pg_appendonly.h:
```diff
@@ -99,29 +99,39 @@ InsertAppendOnlyEntry(Oid relid,
+						  blkdirrelid_ptr, \
+						  visimaprelid_ptr) \
+do { \
+	Form_pg_appendonly aoForm = ((Relation)rel)->rd_appendonly; \
```
We don't need the cast to (Relation).

huansong replied:
Yep, done!

soumyadeep2007 reviewed (approved):
Good change! Feel free to merge once it passes the pipeline (and ensure there are no compiler warnings since some of the type safety goes away with macros).

huansong reviewed (commented):
@soumyadeep2007 Thanks for the quick +1!

> and ensure there are no compiler warnings since some of the type safety goes away with macros

Yep ensured!

ashwinstar commented:
@huansong would be good to have this simple PR committed.

huansong commented on src/include/catalog/pg_appendonly.h:
```diff
@@ -99,29 +99,39 @@ InsertAppendOnlyEntry(Oid relid,
-						  Oid *segrelid,
-						  Oid *blkdirrelid,
-						  Oid *visimaprelid);
+#define GetAppendOnlyEntryAuxOids(rel, \
```
Not sure either but my guess is that it's just an oversight. The original change that introduced `rd_appendonly` to relcache only touched `GetAppendOnlyEntry`: https://github.com/greenplum-db/gpdb/commit/ef92bcf151f99c9152a10c8750eee0d4dde39bc1.

huansong commented:
Sorry I missed this PR. Rebased and addressed additional comments. Will merge after green pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/pgao-macros
